pexec/managed_process_test.go

```diff
@@ -193,6 +209,63 @@ func TestManagedProcessManage(t *testing.T) {
 			test.That(t, err.Error(), test.ShouldContainSubstring, "exit status 1")
 		}
 	})
+	t.Run("OnCrashHandler", func(t *testing.T) {
+		t.Run("returns true and process is restarted", func(t *testing.T) {
+			logger := golog.NewTestLogger(t)
+
+			tempFile := testutils.TempFile(t, "something.txt")
+			defer tempFile.Close()
+
+			watcher, err := fsnotify.NewWatcher()
+			test.That(t, err, test.ShouldBeNil)
+			defer watcher.Close()
+			watcher.Add(tempFile.Name())
+
+			var onCrashHandlerCallCount int
+			proc := NewManagedProcess(ProcessConfig{
+				Name:           "bash",
+				Args:           []string{"-c", fmt.Sprintf("echo hello >> '%s'\nexit 1", tempFile.Name())},
+				OnCrashHandler: func() bool { onCrashHandlerCallCount++; return true },
+			}, logger)
+			test.That(t, proc.Start(context.Background()), test.ShouldBeNil)
+
+			<-watcher.Events
+			<-watcher.Events
+			<-watcher.Events
+
+			// OnCrashHandler should be called twice, as program should crash twice:
+			// first on initial crash, and second on crash after restart.
+			test.That(t, onCrashHandlerCallCount, test.ShouldEqual, 2)
+
+			err = proc.Stop()
+			// sometimes we simply cannot get the status
+			if err != nil {
+				test.That(t, err.Error(), test.ShouldContainSubstring, "exit status 1")
+			}
+		})
+		t.Run("returns false and process is not restarted", func(t *testing.T) {
+			logger := golog.NewTestLogger(t)
+
+			var onCrashHandlerCallCount int
+			proc := NewManagedProcess(ProcessConfig{
+				Name:           "bash",
+				Args:           []string{"-c", "exit 1"},
+				OnCrashHandler: func() bool { onCrashHandlerCallCount++; return false },
+			}, logger)
+			test.That(t, proc.Start(context.Background()), test.ShouldBeNil)
+
+			// OnCrashHandler should be called once, as program should crash once and
+			// not restart.
+			time.Sleep(2 * time.Second)
```

Comment:
This test is bad, and would love any other ideas on better ways to write it. 

The trouble is I'm trying to assert that `onCrashHandler` is called once _and only_ once when it returns false. There's nothing to "wait" for here that will tell me once the process has actually finished. Using a strategy like that of the test above (waiting for writes to a file) does not work: even after the write is awaited (`<-watcher.Events`), the `onCrashHandler` will not have been called yet. The `Sleep` "works" in that it tests that `onCrashHandlerCount` is called once within the two second window, but it's racy and doesn't really test that `onCrashHandlerCount` will be called _only_ once.

Reply:
I don't think this is TOO bad... a two second sleep isn't the end of the world here.

That said, I would consider a dynamic function in the test per my suggestion above. Maybe combine both these tests... have it restart five times, then sleep a couple seconds, and verify that the count is EXACTLY five.

Reply:
> Maybe combine both these tests... have it restart five times, then sleep a couple seconds, and verify that the count is EXACTLY five.

Did exactly this with some safer concurrency.